c7n/actions.py

```diff
@@ -103,53 +103,165 @@ def _run_api(self, cmd, *args, **kw):
+        if 'isolation-group' in self.data:
+            if isinstance(self.data['isolation-group'], basestring) and 'sg-' not in self.data['isolation-group']:
+                raise ValueError('Must specify a valid security group id for the \'isolation-group\' parameter')
+            if isinstance(self.data['isolation-group'], list) and any('sg-' not in g for g in self.data['isolation-group']):
+                raise ValueError('Must specify valid security group ids for the \'isolation-group\' parameter')
+
+        return self
+
+    def get_groups(self, resources, metadata_key=None):
+        """ Parse policies to get lists of security groups to attach to each input resource
+
+        For each input resource, parse the various add/remove/isolation-
+        group policies for 'modify-security-groups' to find the resulting
+        set of VPC security groups to attach to that resource.
+
+        The 'metadata_key' parameter can be used for two purposes at the moment;
+        The first use is for resources' APIs that return a list of security group
+        IDs but use a different metadata key than 'Groups' or 'SecurityGroups'.
+
+        The second use is for when there are richer objects in the 'Groups' or
+        'SecurityGroups' lists. The custodian actions need to act on lists of
+        just security group IDs, so the metadata_key can be used to select IDs
+        from the richer objects in the provided lists.
+
+        Returns a list of lists containing the resulting VPC security groups
+        that should end up on each resource passed in.
+
+        :param resources: List of resources containing VPC Security Groups
+        :param metadata_key: Metadata key for security groups list (see notes for detailed usage)
+        :return: List of lists of security groups per resource
+        """
+        # parse the add, remove, and isolation group params to return the
+        # list of security groups that will end up on the resource
+        # target_group_ids = self.data.get('groups', 'matched')
+
+        add_target_group_ids = self.data.get('add', None)
+        remove_target_group_ids = self.data.get('remove', None)
         isolation_group = self.data.get('isolation-group')
+        add_groups = []
+        remove_groups = []
         return_groups = []
 
-        for r in resources:
-            rgroups = [g['GroupId'] for g in r['Groups']]
-            if target_group_ids == 'matched':
-                group_ids = r.get('c7n.matched-security-groups', ())
-            elif target_group_ids == 'all':
-                group_ids = rgroups
-            elif isinstance(target_group_ids, list):
-                group_ids = target_group_ids
-            else:
-                continue
-
-            if not group_ids:
-                continue
-
-            for g in group_ids:
+        for idx, r in enumerate(resources):
+            if r.get('Groups'):
```

Comment:
sans adaptation, this retrieval of sg ids from a resource belongs in the subclasses not the base class. a get_resource_sgs(resource) method would suffice.